decimal/decimal_test.go

```diff
@@ -0,0 +1,574 @@
+	}
+	return conn
+}
+
+func tupleValueIsDecimal(t *testing.T, tuples []interface{}, number decimal.Decimal) {
+	if len(tuples) != 1 {
+		t.Fatalf("Response Data len (%d) != 1", len(tuples))
+	}
+
+	if tpl, ok := tuples[0].([]interface{}); !ok {
+		t.Fatalf("Unexpected return value body")
+	} else {
+		if len(tpl) != 1 {
+			t.Fatalf("Unexpected return value body (tuple len)")
+		}
+		if val, ok := tpl[0].(decimal.Decimal); !ok || !val.Equal(number) {
+			t.Fatalf("Unexpected return value body (tuple 0 field)")
+		}
+	}
+}
+
+func trimMPHeader(mpBuf []byte, fixExt bool) []byte {
+	mpHeaderLen := 2
+	if fixExt == false {
+		mpHeaderLen = 3
+	}
+	return mpBuf[mpHeaderLen:]
+}
+
+func TestEncodeStringToBCD(t *testing.T) {
+	samples := append(correctnessSamples, benchmarkSamples...)
+	for _, testcase := range samples {
+		t.Run(testcase.numString, func(t *testing.T) {
+			buf, err := EncodeStringToBCD(testcase.numString)
+			if err != nil {
+				t.Fatalf("Failed to encode decimal '%s' to BCD: %s", testcase.numString, err)
+
+			}
+			b, _ := hex.DecodeString(testcase.mpBuf)
+			bcdBuf := trimMPHeader(b, testcase.fixExt)
+			if reflect.DeepEqual(buf, bcdBuf) != true {
+				t.Fatalf("Failed to encode decimal '%s' to BCD: expected '%x', actual '%x'", testcase.numString, bcdBuf, buf)
+			}
+		})
+	}
+}
+
+func TestDecodeStringFromBCD(t *testing.T) {
+	samples := append(correctnessSamples, benchmarkSamples...)
+	for _, testcase := range samples {
+		t.Run(testcase.numString, func(t *testing.T) {
+			b, _ := hex.DecodeString(testcase.mpBuf)
+			bcdBuf := trimMPHeader(b, testcase.fixExt)
+			s, err := DecodeStringFromBCD(bcdBuf)
+			if err != nil {
+				t.Fatalf("Failed to decode BCD '%x' to decimal: %s", bcdBuf, err)
+			}
+
+			decActual, err := decimal.NewFromString(s)
+			if err != nil {
+				t.Fatalf("Failed to encode string ('%s') to decimal", s)
+			}
+			decExpected, err := decimal.NewFromString(testcase.numString)
+			if err != nil {
+				t.Fatalf("Failed to encode string ('%s') to decimal", testcase.numString)
+			}
+			if !decExpected.Equal(decActual) {
+				t.Fatalf("Decoded decimal from BCD ('%x') is incorrect: expected '%s', actual '%s'", bcdBuf, testcase.numString, s)
+			}
+		})
+	}
+}
+
+func BenchmarkEncodeStringToBCD(b *testing.B) {
+	for _, testcase := range benchmarkSamples {
+		b.Run(testcase.numString, func(b *testing.B) {
+			b.ResetTimer()
+			for n := 0; n < b.N; n++ {
+				EncodeStringToBCD(testcase.numString)
+			}
+		})
+	}
+}
+
+func BenchmarkDecodeStringFromBCD(b *testing.B) {
+	for _, testcase := range benchmarkSamples {
+		b.Run(testcase.numString, func(b *testing.B) {
+			buf, _ := hex.DecodeString(testcase.mpBuf)
+			bcdBuf := trimMPHeader(buf, testcase.fixExt)
+			b.ResetTimer()
+			for n := 0; n < b.N; n++ {
+				DecodeStringFromBCD(bcdBuf)
+			}
+		})
+	}
+}
+
+func TestSelect(t *testing.T) {
+	if isDecimalSupported == false {
+		t.Skip("Skipping test for Tarantool without decimal support in msgpack")
+	}
```

Comment:
The code can be extracted into a separate test helper function.

Reply:
Could you explain what for? What's a problem with this code?

Reply:
You can extract a separate helper function for skip and not repeat the check before each. This is not very critical, in fact, it will just replace 3 lines with 1. So you can ignore it.
```
func skipIfDatetimeUnsupported(t testing.TB) {
    t.Helper()
    if isDecimalSupported == false {
        t.Skip("Skipping test for Tarantool without datetime support in msgpack")
    }
}
// ...
func TestCaseAnyCaseThatWeWantToSkip(t *testing.T) {
    skipIfDecimalUnsupported(t)
    // ...
}
```